### PR TITLE
Feature/page requests

### PR DESCRIPTION
**Pull request by rhyeen**
Added endpoints for:
* GET pages (list of pages, paginated)
* GET page
* GET entire page
* DELETE page (sets page to "deletedAt", but doesn't actually delete)
* PATCH page (set page)
* POST page (create page)

Also added:
* GUID generator for easily generating new guids with future POST detail/property/relation/etc. requests.
* Query helpers for making handling of the SQL driver easier for basic calls such as insert, update, and select one row.
* SQL query generators for easily creating SQL queries like an ORM but with less frameworky coupling.

### CODE REVIEW

File: internal/api/handlers/page/request.go [hunk 1]
```diff
@@ -32,9 +33,12 @@ func (request CreatePageRequest) validate() (CreatePageRequest, error) {
 	if request.Title == "" {
 		return request, errors.New("must provide title")
 	}
-	if request.VersionID <= 0 {
+	if request.VersionID == "" {
 		return request, errors.New("must provide versionId")
 	}
+	if request.PageTemplateID == "" {
+		return request, errors.New("msut provide pageTemplateId")
```
Comment (Pergamene) [inline internal/api/handlers/page/request.go]: Fix "msut"

File: internal/api/handlers/page/handler_test.go [hunk 2]
```diff
@@ -139,7 +144,7 @@ func TestUpdatePage(t *testing.T) {
 		datacenter           string
 		expectedResponseBody string
 		expectedStatusCode   int
-		updatePageCalls      []updatePageCall
+		updatePageCalls         []updatePageCall
```
Comment (Pergamene) [inline internal/api/handlers/page/handler_test.go]: Is the extra white space in front of `[]updatePageCall` intentional? How did you even do that with go?

File: internal/services/page/service_test.go [hunk 3]
```diff
@@ -50,40 +42,105 @@ func getStoreUnauthorizedErr(userID, tableID string, err error) error {
...
-		params           UpdatePageParams
-		canEditPageCalls []canEditPageCall
-		updatePageCalls  []updatePageCalls
-		returnErr        error
+		name                 string
+		params               UpdatePageParams
+		canEditPageCalls     []canEditPageCall
+		getPageTemplateCalls []getPageTemplateCall
+		getVersionCalls      []getVersionCall
+		updatePageCalls         []updatePageCall
```
Comment (Pergamene) [inline internal/services/page/service_test.go]: Inconsistant spacing before `[]updatePageCall`

File: internal/stores/mysqlstore/healthcheckstore_test.go [hunk 4]
```diff
@@ -33,15 +32,15 @@ func TestHealthcheckIsHealthy(t *testing.T) {
 			returnIsHealthy: false,
 		},
 		{
-			name:                   "db not setup",
+			name: "db not setup",
```
Comment (Pergamene) [inline internal/stores/mysqlstore/healthcheckstore_test.go:35]: Shouldn't "db not setup", be indented like the rest?

File: internal/stores/mysqlstore/pagestore.go [hunk 5]
```diff
@@ -27,192 +27,354 @@ func NewPageStore(mysqldb *sql.DB) PageStore {
...
-	}
 	if s.db == nil {
 		return isOwner, &storeerror.DBNotSetUp{}
 	}
-	rows, err := s.db.Query("SELECT `permission` FROM `Page` WHERE `guid` = ?", pageGUID)
-	if err != nil {
-		return isOwner, err
+	statement := wrapsql.SelectStatement{
+		Selectors: []string{"permission"},
+		FromTable: "Page",
```
Comment (Pergamene) [inline internal/stores/mysqlstore/pagestore.go:142]: Shouldn't `[]string{"permission"},` and `"Page",` be indented like the next line?

File: internal/stores/mysqlstore/pagestore.go [hunk 6]
```diff
@@ -27,192 +27,354 @@ func NewPageStore(mysqldb *sql.DB) PageStore {
...
+	return total, nil
+}
+
+func (s PageStore) getPageID(guid string) (int, error) {
+	if guid == "" {
+		return -1, errors.New("must provide guid to get the page id")
+	}
+	statement := wrapsql.SelectStatement{
+		Selectors: []string{"ID"},
+		FromTable: "Page",
```
Comment (Pergamene) [inline internal/stores/mysqlstore/pagestore.go:345]: Indenting?

File: internal/stores/mysqlstore/pagestore_test.go [hunk 7]
```diff
@@ -72,3 +288,285 @@ func TestPageUpdatePage(t *testing.T) {
...
+				"INSERT INTO Page (`Version_ID`, `PageTemplate_ID`, `guid`, `title`, `summary`, `permission`, `createdAt`, `updatedAt`) VALUES( 1, 1, \"PG_4\", \"test title 4\", \"\", \"PR\", NOW(), NOW() )",
+				"INSERT INTO PageOwner (`Page_ID`, `User_ID`, `isOwner`) VALUES( 1, 1, true)",
+				"INSERT INTO PageOwner (`Page_ID`, `User_ID`, `isOwner`) VALUES( 2, 1, true)",
+				"INSERT INTO PageOwner (`Page_ID`, `User_ID`, `isOwner`) VALUES( 3, 1, true)",
+				"INSERT INTO PageOwner (`Page_ID`, `User_ID`, `isOwner`) VALUES( 4, 2, true)",
+			},
+			paramUserID:      "UR_1",
+			paramThisBatchID: "PG_3",
+			paramLimit:       2,
+			returnPages: []page.Page{
```
Comment (Pergamene) [inline internal/stores/mysqlstore/pagestore_test.go:422]: indenting?

File: internal/util/wrapsql/statementbuilder_test.go [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+		name                 string
+		paramSelectStatement SelectStatement
+		returnStatement      string
+	}{
+		{
+			name: "test 'get page' statement",
+			paramSelectStatement: SelectStatement{
+				Selectors: []string{"Page.ID", "Version.guid", "PageTemplate.guid", "Page.title", "Page.summary", "Page.permission", "Page.createdAt", "Page.updatedAt"},
+				FromTable: "Page",
+				JoinClauses: []JoinClause{
```
Comment (Pergamene) [inline internal/util/wrapsql/statementbuilder_test.go:28]: indenting?

File: static/docs/src/openapi.yaml [hunk 9]
```diff
@@ -283,12 +283,13 @@ paths:
...
                 $ref: '#/definitions/meta'
-    put:
+    patch:
       tags:
       - page
-      summary: Set Page
-      description: Sets the provided page.
-      operationId: setPage
+      summary: Update Page
+      description: |
```
Comment (Pergamene) [inline static/docs/src/openapi.yaml:290]: Is the `|` just so you could put the description all on the next line?

File: static/docs/src/pages.yaml [hunk 10]
```diff
@@ -114,8 +138,24 @@ definitions:
...
+    type: string
+    enum:
+    - PR
+    - PU
+    - PO
+    - LO
+    description: |
+      * **PR**: Private. Only the owner(s) may edit/see the page.
+      * **PU**: Public.  Everyone may edit/see the page and the page is searchable.
+      * **PO**: Public Only. Everyone may see the page and the page is searchable.
```
Comment (Pergamene) [inline static/docs/src/pages.yaml:157]: When would you do `PO` over `PU`?

**Review by Pergamene — CHANGES_REQUESTED** [review]
1 typo, 1 question, the rest I think were just weird indenting.

File: static/docs/src/pages.yaml (hunk 10 above)
Comment (rhyeen) [reply static/docs/src/pages.yaml:157]: PO: If you want to make something publicly viewable that only you want edit access to
PU: If you want to make something that anyone can modify, like a wiki page

File: static/docs/src/openapi.yaml (hunk 9 above)
Comment (rhyeen) [reply static/docs/src/openapi.yaml:290]: Right.  It's especially helpful if the description is going to be multiline

File: internal/stores/mysqlstore/healthcheckstore_test.go (hunk 4 above)
Comment (rhyeen) [reply internal/stores/mysqlstore/healthcheckstore_test.go:35]: This one doesn't want to update...so I think it's right.  No idea why though.

File: internal/stores/mysqlstore/pagestore.go (hunk 5 above)
Comment (rhyeen) [reply internal/stores/mysqlstore/pagestore.go:142]: Anytime there is an `[]X{}` or `X{}` with actual content in the `{}`, it starts a new indent spacing format

File: internal/stores/mysqlstore/pagestore.go (hunk 6 above)
Comment (rhyeen) [reply internal/stores/mysqlstore/pagestore.go:345]: same as above

File: internal/stores/mysqlstore/pagestore_test.go (hunk 7 above)
Comment (rhyeen) [reply internal/stores/mysqlstore/pagestore_test.go:422]: same

File: internal/util/wrapsql/statementbuilder_test.go (hunk 8 above)
Comment (rhyeen) [reply internal/util/wrapsql/statementbuilder_test.go:28]: same

**Pull request merged**